source/shared/cpp/ObjectModel/ParseUtil.cpp

```diff
@@ -150,29 +150,34 @@ namespace AdaptiveSharedNamespace
 
     std::shared_ptr<BackgroundImage> ParseUtil::GetBackgroundImage(const Json::Value& json)
     {
-        try
-        {
-            // handle "backgroundImage": <string>
-            std::string backgroundImageUrl = ParseUtil::GetString(json, AdaptiveCardSchemaKey::BackgroundImage, false);
-            if (backgroundImageUrl != "")
-            {
-                return std::make_shared<BackgroundImage>(backgroundImageUrl);
-            }
+        // handle "backgroundImage": <string>
+        const std::string& backgroundImagePropertyName = AdaptiveCardSchemaKeyToString(AdaptiveCardSchemaKey::BackgroundImage);
```

Comment:
Curious about the guideline/best practice here -- is it that we should avoid exception code paths int eh "golden" path scenarios but that it's ok to report back errors directly from here via exceptions in the error scenarios (eg GetValueAsString above that throws)

Reply:
Best practice is that exceptions shouldn't be used for implementing behaviour that is normal and expected (in this case a backgroundImage property that has an object value). In isolation it's fine for ParseUtil::GetString to throw but using such a function in GetBackgroundImage, when we expect to be be passing the function an 'invalid' argument, isn't a good idea. It's the wrong tool for the job, so to speak. The [CppCoreGuidelines](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Re-throw) (E2 & E3) do a good job of explaining this.

On a practical level I want to set the debugger to break when all exceptions are thrown because it allows me to discover potential bugs - I know that if any exception is thrown during development or testing that something unexpected has happened and I should immediately investigate. It's not really possible to do this if the software is constantly generating exceptions.